### ISSUE TITLE
Proper types for link context

### ISSUE THREAD

**Issue by ab-pm** [apollographql/apollo-client#9249]
Closely related to #8902 but with a larger scope.

The type of the `context` that is passed from a query through to the links is currently `Record<string, any>` or even worse, just `any`.

https://github.com/apollographql/apollo-client/blob/a4b03cdb69ffb5164f0db2e34f842307ab95dba4/src/core/watchQueryOptions.ts#L72-L75
https://github.com/apollographql/apollo-client/blob/b214dd1d31cdfe6b5477bb5887a7d9615a302b27/src/link/core/types.ts#L10
https://github.com/apollographql/apollo-client/blob/b214dd1d31cdfe6b5477bb5887a7d9615a302b27/src/link/core/types.ts#L19-L20
https://github.com/apollographql/apollo-client/blob/b0da267d24d8713987e9ff8d2240c5c7ed94eb94/src/core/types.ts#L16
https://github.com/apollographql/apollo-client/blob/b214dd1d31cdfe6b5477bb5887a7d9615a302b27/src/link/context/index.ts#L4-L7

`FetchResult` already is generic, but does not supply a type parameter for `TContext` when used anywhere, making it fallback to the default
https://github.com/apollographql/apollo-client/blob/b214dd1d31cdfe6b5477bb5887a7d9615a302b27/src/link/core/types.ts#L23-L31

I request (and would be **happy to supply a PR** with) two main fixes:

* use the `DefaultContext` type (that is also exported as `Context` from `@apollo/client`) consistently in the codebase, instead of duplicating `Record<string, any>`/`any`
* change the `type` alias into an interface. This will allow [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) so users of the library can declare their own extensions of the context type, and get proper types and intellisense annotations when accessing properties on the context. This includes properties like `headers` for the native `HttpLink` [context options](https://www.apollographql.com/docs/react/api/link/apollo-link-http/#context-options), properties from plugins like `RestLink` (which already declare [their own `LinkChainContext`](https://github.com/apollographql/apollo-link-rest/blob/23edd85b2769273c9c59c6615ef95ff598708349/src/restLink.ts#L770-L786)), and entirely custom properties that e.g. can be used to [`split` a link](https://www.apollographql.com/docs/react/api/link/introduction/#directional-composition).
  ```ts
  export interface DefaultContext {
    /** @deprecated use an explicitly declared property instead */
    [p: string]: any; 
  }
  ```
  I would prefer the interface to be empty (and only be amended by other modules that actually use specific properties), but the index signature is required for backward compatibility. It might be removed in a major release.

For real type safety, it would of course be nice if one could make `ApolloClient`, `QueryManager` and `ApolloLink` generic over the context (the link getting two type parameters, one for upstream context type and one for the downstream context type), so that the `concat` method could ensure the right context type is used upstream and that methods like `query` and `mutate` expect context values fitting into the link that the client uses. But that would probably be a major breaking change, and the extra complexity is not worth the marginal gains. Most applications don't use that many different context types…

**Comment by philipbulley** [comment]
The `Context` interface with declaration merging seems like an easy "low hanging fruit" approach to achieving this 👍🏽

**Comment by shaunbaruth** [comment]
Our team would love to see this implemented as well 👍

**Comment by Hugodby** [comment]
Hello @ab-pm, I submitted the following PR  #10402 based on your suggestion

**Comment by alessbell** [comment]
Thanks to @Hugodby for the PR! It's just been merged and will go out in the next release, `3.7.5`, some time next week :)

**Comment by chr-ge** [comment]
What's the right way of merging the `DefaultContext` type? Doing it this way breaks the `@apollo/client` types:
```ts
// apollo.d.ts

declare module '@apollo/client' {
  interface DefaultContext {
    ...
  }
}
```

**Comment by Hugodby** [comment]
Hello @chr-ge,
You can merge it like this:
```
// apollo.d.ts

import '@apollo/client';

declare module '@apollo/client' {
  interface DefaultContext {
    ...
  }
}
```

**Comment by chr-ge** [comment]
> Hello @chr-ge, You can merge it like this:
> 
> ```ts
> // apollo.d.ts
> 
> import '@apollo/client';
> 
> declare module '@apollo/client' {
>   interface DefaultContext {
>     ...
>   }
> }
> ```

Silly me, thanks

**Comment by jerelmiller** [comment]
This issue was fixed with #10402 and released with 3.7.5. Thanks!
